app/src/main/java/com/islaharper/dawnofandroid/presentation/screens/welcome/WelcomeScreen.kt

```diff
@@ -1,12 +1,176 @@
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.foundation.layout.padding
+import androidx.compose.material3.Button
+import androidx.compose.material3.ButtonDefaults
+import androidx.compose.material3.MaterialTheme
+import androidx.compose.material3.Text
 import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.font.FontWeight
+import androidx.compose.ui.text.style.TextAlign
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.hilt.navigation.compose.hiltViewModel
 import androidx.navigation.NavHostController
+import com.google.accompanist.pager.ExperimentalPagerApi
+import com.google.accompanist.pager.HorizontalPager
+import com.google.accompanist.pager.HorizontalPagerIndicator
+import com.google.accompanist.pager.PagerState
+import com.google.accompanist.pager.rememberPagerState
+import com.islaharper.dawnofandroid.R
+import com.islaharper.dawnofandroid.domain.model.OnBoardingPage
+import com.islaharper.dawnofandroid.navigation.Screen
+import com.islaharper.dawnofandroid.ui.theme.AndroidGreen
+import com.islaharper.dawnofandroid.ui.theme.Neutral90
+import com.islaharper.dawnofandroid.ui.theme.PADDING_SMALL
+import com.islaharper.dawnofandroid.ui.theme.PADDING_XLARGE
+import com.islaharper.dawnofandroid.util.Constants.ONBOARDING_LAST_PAGE
+import com.islaharper.dawnofandroid.util.Constants.ONBOARDING_PAGE_COUNT
 
+@OptIn(ExperimentalPagerApi::class)
 @Composable
 fun WelcomeScreen(
     navHostController: NavHostController,
+    welcomeViewModel: WelcomeViewModel = hiltViewModel(),
 ) {
-    Text(text = "Welcome Placeholder")
+    val pages = listOf(
+        OnBoardingPage.FirstPage,
+        OnBoardingPage.SecondPage,
+        OnBoardingPage.ThirdPage,
+    )
+    val pagerState = rememberPagerState()
+
+    Column(
+        modifier = Modifier
+            .fillMaxSize()
+            .background(color = MaterialTheme.colorScheme.background),
+    ) {
+        HorizontalPager(
+            modifier = Modifier.weight(10f),
+            count = ONBOARDING_PAGE_COUNT,
+            state = pagerState,
+            verticalAlignment = Alignment.Top,
+        ) { position ->
+            PagerScreen(page = pages[position])
+        }
+        HorizontalPagerIndicator(
+            modifier = Modifier.weight(1f).align(Alignment.CenterHorizontally),
+            pagerState = pagerState,
+            activeColor = AndroidGreen,
+            inactiveColor = Neutral90,
+            indicatorWidth = 12.dp,
+            spacing = 8.dp,
+        )
+        FinishButton(
+            modifier = Modifier.weight(1f).padding(horizontal = PADDING_XLARGE),
+            pagerState = pagerState,
+        ) {
+            welcomeViewModel.saveOnBoardingState(completed = true)
```

Comment:
Looks like the `FinishButton` is using the `ViewModel` here. Passing down the ViewModel to lower level composables is something we probably want to avoid because the ViewModel is not stable.

I think we should only use the ViewModel at the top level composable, and pass down lambdas.

So the top level composable (the screen) should have the ViewModel as a parameter (through hilt). Then pass down lambdas to the other Compsables:

```kotlin
    val onFinishClick = { welcomeViewModel.saveOnBoardingState(true) }

        FinishButton(
            modifier = Modifier.weight(1f).padding(horizontal = PADDING_XLARGE),
            pagerState = pagerState,
            onFinishClick,
        )
```

Can you have a look at the Twitter Compose rules because they advocate for something like this, I think?

https://slackhq.github.io/compose-lints/

Reply:
Change made in latest commit.